pkg/providers/cloudstack/validator.go

```diff
@@ -0,0 +1,197 @@
+	}
+	logger.MarkPass("Connected to server")
+
+	return nil
+}
+
+func (v *Validator) ValidateCloudStackDatacenterConfig(ctx context.Context, datacenterConfig *anywherev1.CloudStackDatacenterConfig) error {
+	if err := v.cmk.ValidateZonePresent(ctx, datacenterConfig.Spec.Zone); err != nil {
+		return err
+	}
+	logger.MarkPass("Datacenter validated")
+
+	return nil
+}
+
+// TODO: dry out machine configs validations
+func (v *Validator) ValidateClusterMachineConfigs(ctx context.Context, clusterSpec *Spec) error {
+	var etcdMachineConfig *anywherev1.CloudStackMachineConfig
+
+	if len(clusterSpec.Cluster.Spec.ControlPlaneConfiguration.Endpoint.Host) <= 0 {
+		return errors.New("cluster controlPlaneConfiguration.Endpoint.Host is not set or is empty")
+	}
+	if len(clusterSpec.datacenterConfig.Spec.Domain) <= 0 {
+		return errors.New("CloudStackDatacenterConfig domain is not set or is empty")
+	}
+	if len(clusterSpec.datacenterConfig.Spec.Zone.Value) <= 0 {
+		return errors.New("CloudStackDatacenterConfig zone is not set or is empty")
+	}
+	if len(clusterSpec.datacenterConfig.Spec.Network.Value) <= 0 {
+		return errors.New("CloudStackDatacenterConfig network is not set or is empty")
+	}
+	if clusterSpec.Cluster.Spec.ControlPlaneConfiguration.MachineGroupRef == nil {
+		return errors.New("must specify machineGroupRef for control plane")
+	}
+	controlPlaneMachineConfig := clusterSpec.controlPlaneMachineConfig()
+
+	if clusterSpec.Cluster.Spec.WorkerNodeGroupConfigurations[0].MachineGroupRef == nil {
+		return errors.New("must specify machineGroupRef for worker nodes")
+	}
+
+	workerNodeGroupMachineConfig, ok := v.machineConfigs[clusterSpec.Cluster.Spec.WorkerNodeGroupConfigurations[0].MachineGroupRef.Name]
+	if !ok {
+		return fmt.Errorf("cannot find CloudStackMachineConfig %v for worker nodes", clusterSpec.Cluster.Spec.WorkerNodeGroupConfigurations[0].MachineGroupRef.Name)
+	}
+
+	if clusterSpec.Cluster.Spec.ExternalEtcdConfiguration != nil {
+		if clusterSpec.Cluster.Spec.ExternalEtcdConfiguration.MachineGroupRef == nil {
+			return errors.New("must specify machineGroupRef for etcd machines")
+		}
+		etcdMachineConfig = clusterSpec.etcdMachineConfig()
+		if len(etcdMachineConfig.Spec.Users) <= 0 {
+			etcdMachineConfig.Spec.Users = []anywherev1.UserConfiguration{{}}
+		}
+		if len(etcdMachineConfig.Spec.Users[0].SshAuthorizedKeys) <= 0 {
+			etcdMachineConfig.Spec.Users[0].SshAuthorizedKeys = []string{""}
+		}
+	}
+
+	if len(controlPlaneMachineConfig.Spec.Users) <= 0 {
+		controlPlaneMachineConfig.Spec.Users = []anywherev1.UserConfiguration{{}}
+	}
+	if len(workerNodeGroupMachineConfig.Spec.Users) <= 0 {
+		workerNodeGroupMachineConfig.Spec.Users = []anywherev1.UserConfiguration{{}}
+	}
+	if len(controlPlaneMachineConfig.Spec.Users[0].SshAuthorizedKeys) <= 0 {
+		controlPlaneMachineConfig.Spec.Users[0].SshAuthorizedKeys = []string{""}
+	}
+	if len(workerNodeGroupMachineConfig.Spec.Users[0].SshAuthorizedKeys) <= 0 {
```

Comment:
It seems like we have a lot of duplicated code for validating the machine configs
Could we form a slice of cloud stack machine configs and iterate over them, running the common validations?

Reply:
Yea we added support for worker node groups when we refactored this a little which doesn't seem to be included here. I forget, we aren't planning on supporting multiple worker node groups from the start right?

Reply:
Correct, that would make the initial merge more complicated, so I'd prefer to keep worker node groups out if possible. I'll see what I can do about iterating/validating, thanks for the tip :)

Reply:
My suggestion was independent from supporting worker node groups or not
There is duplicated code here when validating machine configs for cp, etcd and workers (it really doesn't matter if we support multiple worker node groups here, since all the worker node groups could be pointing to the same cloud stack machine config).
All of these machine configs can be grouped in a slice and run the common validations from there

As far as aligning with other provider patterns, I would suggest to make a judgment call case by case. Not all the code present for other providers is ideal. Most of it has been built over time and the patterns that made sense in the past might not make sense anymore.  We should refactor them if they don't, of course. But that's a different topic. 

I think we should write this new code in the way that makes the most sense now.

Reply:
Refactored to perform more checks inside the machineConfig for loop, while still trying to keep some valuable error messaging in place